crates/core/src/subscription/module_subscription_actor.rs

```diff
@@ -185,45 +181,44 @@ impl ModuleSubscriptions {
         &self,
         client: Option<&ClientConnectionSender>,
         mut event: ModuleEvent,
-        ctx: &ExecutionContext,
         tx: MutTx,
     ) -> Result<Result<Arc<ModuleEvent>, WriteConflict>, DBError> {
         // Take a read lock on `subscriptions` before committing tx
         // else it can result in subscriber receiving duplicate updates.
         let subscriptions = self.subscriptions.read();
         let stdb = &self.relational_db;
-
+        let ctx = tx.ctx.clone();
         // Downgrade mutable tx.
         // Ensure tx is released/cleaned up once out of scope.
-        let read_tx = scopeguard::guard(
+        let mut read_tx = scopeguard::guard(
             match &mut event.status {
                 EventStatus::Committed(db_update) => {
-                    let Some((tx_data, read_tx)) = stdb.commit_tx_downgrade(ctx, tx)? else {
+                    let Some((tx_data, read_tx)) = stdb.commit_tx_downgrade(tx)? else {
                         return Ok(Err(WriteConflict));
                     };
                     *db_update = DatabaseUpdate::from_writes(&tx_data);
                     read_tx
                 }
-                EventStatus::Failed(_) | EventStatus::OutOfEnergy => stdb.rollback_mut_tx_downgrade(ctx, tx),
+                EventStatus::Failed(_) | EventStatus::OutOfEnergy => stdb.rollback_mut_tx_downgrade(tx),
             },
             |tx| {
-                self.relational_db.release_tx(ctx, tx);
+                self.relational_db.release_tx(tx);
             },
         );
         let event = Arc::new(event);
 
         // New execution context for the incremental subscription update.
         // TODO: The tx and the ExecutionContext should be coupled together.
-        let ctx = if let Some(reducer_ctx) = ctx.reducer_context() {
+        read_tx.ctx = if let Some(reducer_ctx) = ctx.reducer_context() {
             ExecutionContext::incremental_update_for_reducer(stdb.address(), reducer_ctx.clone())
         } else {
             ExecutionContext::incremental_update(stdb.address())
         };
```

Comment:
This entire `if let` block can be removed since `read_tx` produced by `commit_tx_downgrade` automatically contains the correct execution context.